Timeout for the CreateBucket call to minio (or S3)

tranchitella opened the pull request:
The deployments service creates the S3 bucket at startup time: both
NewSimpleStorageServiceStatic and NewSimpleStorageServiceDefaults call
`client.CreateBucket` to initialize the bucket.

In case of error, the previous implementation returned `nil` and this caused
the process to exit with exit code `4`.

This commit applies a timeout (using `context.WithTimeout`) set to 5 seconds in
order to avoid the deployments service to hang and wait an indefinite amount of
time before retrying the command.

Changelog: title
Signed-off-by: Fabio Tranchitella <fabio@tranchitella.eu>

tranchitella commented on s3/filestorage.go:
```diff
@@ -89,14 +89,30 @@ func NewSimpleStorageServiceStatic(bucket, key, secret, region, token, uri strin
+		} else {
+			break
 		}
+		l.Warnf(fmt.Sprintf("Unable to initialize the S3 bucket, sleeping 1 second before retrying [%d]...", i))
```
Thank you for your feedback!

kjaskiewiczz commented:
if we are assuming that minio should be operational when the deployments service is starting then it's ok for the deployments service to exit if it cannot connect to the minio
same thing as with connecting to mongo - https://github.com/mendersoftware/deployments/blob/c1e7da299f8843e52744974db304be630d54bba4/main.go#L142

tranchitella commented:
@kjaskiewiczz 

This PR still causes deployments to exit if minio is not ready, it just implements a retry-logic if minio is up and running but not ready yet to accept requests (as it opens the TCP port before being ready, and replies to API calls with 5xx) and introduces a timeout to avoid deployments hanging too long waiting for an answer.

The latter is what we experience in the integration test pipeline: when the containers get restarted by the `restart_docker_compose` method it can happen deployments service resolves `s3.docker.mender.io` with the wrong IP address and waits more than 30 seconds before dying.

mchalski commented:
@tranchitella 

TBH I'm also not a fan of retries, especially if they're 'for tests' only.

since we're distributed, every service can come and go at any moment. take an example of a failed db operation - I'm immediately returning 500 because the service simply can't do its job. it doesn't matter if it's in an aws region that died, or it's overwhelmed, or being redeployed, restarted etc. one of the cases is - it's just being started up and still not ready to serve. 

I wouldn't handle it differently here, just die and restart and maybe we're luckier next time. I'd go towards a centralized health check for minio instead (I know we tried, and not sure what's the status).

a sane timeout is a good idea though. 

btw what's the issue with wrong s3 resolution? because this sounds worrying.

tranchitella commented:
@mchalski 

As discussed last week with @kjaskiewiczz I disagree and I expect micro-services to implement a (temporarily limited) retry-logic to wait for the services they depend on. This is how it works in the Kubernetes world where you cannot express dependencies between the PODs and where having PODs that restart several times in a raw can cause them to not be scheduled anymore as they enter a state called `crashloopbackoff`.

Anyway, as I think the timeout is needed anyway I've reworked the PR to contain only that specific change. Can you please review and, if it looks okay, merge it into master?

Regarding your question, the wrong s3 resolution is linked to the specific use case we implemented in the integration tests. We restart the docker containers (aka remove and start them again) and this causes a new IP address to be assigned to them. If the deployments container starts before the minio one (which is not the case now as we implemented the dependency using docker-compose version 2.1) it can try to call the CreateBucket on the old ip address which is not available anymore. In this case the default timeout applies which is quite high (>30 seconds) causing the deployments container to be up and running but without binding the 8080 port (as it hasn't reached yet the ListenAndServe).

This is the reason why we experienced the connection refused / too many 502 errors issue.
